tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,769 @@
+
+	Dim(n int) int
+	NumDims() int
+
+	Bool() (res []bool, err error)
+	Float32() (res []float32, err error)
+	Float64() (res []float64, err error)
+	Int32() (res []int32, err error)
+	Int64() (res []int64, err error)
+	Str() (res [][]byte, err error)
+
+	String() string
+}
+
+// A Tensor holds a multi-dimensional array of elements of a single data type.
+type Tensor struct {
+	pb.TensorProto
+
+	tensor      TF_Tensor
+	dimWeights  []int
+	memReleased bool
+}
+
+// A TensorShape represents the shape of a Tensor.
+type TensorShape [][]int64
+
+var (
+	// DTInvalid Invalid tensor DataType.
+	DTInvalid = DataType(0)
+	// DTBfloat corresponds to TF_BFLOAT16.
+	DTBfloat = DataType(TF_BFLOAT16)
+	// DTBool corresponds to TF_BOOL.
+	DTBool = DataType(TF_BOOL)
+	// DTComplex corresponds to TF_COMPLEX.
+	DTComplex = DataType(TF_COMPLEX)
+	// DTFloat corresponds to TF_FLOAT.
+	DTFloat = DataType(TF_FLOAT)
+	// DTDouble corresponds to TF_DOUBLE.
+	DTDouble = DataType(TF_DOUBLE)
+	// DTInt8 corresponds to TF_INT8.
+	DTInt8 = DataType(TF_INT8)
+	// DTInt16 corresponds to TF_INT16.
+	DTInt16 = DataType(TF_INT16)
+	// DTInt32 corresponds to TF_INT32.
+	DTInt32 = DataType(TF_INT32)
+	// DTInt64 corresponds to TF_INT64.
+	DTInt64 = DataType(TF_INT64)
+	// DTQint16 corresponds to TF_QINT16.
+	DTQint16 = DataType(TF_QINT16)
+	// DTQuint16 corresponds to TF_QUINT16.
+	DTQuint16 = DataType(TF_QUINT16)
+	// DTQuint32 corresponds to TF_QINT32.
+	DTQuint32 = DataType(TF_QINT32)
+	// DTQint8 corresponds to TF_QINT8.
+	DTQint8 = DataType(TF_QINT8)
+	// DTQuint8 corresponds to TF_QUINT8.
+	DTQuint8 = DataType(TF_QUINT8)
+	// DTString corresponds to TF_STRING.
+	DTString = DataType(TF_STRING)
+	// DTUint8 corresponds to TF_UINT8.
+	DTUint8 = DataType(TF_UINT8)
+	// DTUint16 corresponds to TF_UINT16.
+	DTUint16 = DataType(TF_UINT16)
+)
+
+// NewTensorWithShape returns a new tensor with the specified type, shape and data.
+// The supported  data types are:
+//  - DTInt8
+//  - DTInt16
+//  - DTInt32
+//  - DTInt64
+//  - DTUint8
+//  - DTUint16
+//  - DTFloat
+//  - DTDouble
+func NewTensorWithShape(shape TensorShape, data interface{}) (*Tensor, error) {
+	v := reflect.ValueOf(data)
+	if v.Kind() != reflect.Slice {
+		return nil, &ErrSliceExpected{
+			dataType: v.Kind().String(),
+		}
+	}
+
+	dataType, err := getDataTypeFromReflect(v.Type().Elem().Kind(), int64(v.Type().Elem().Size()))
+	if err != nil {
+		return nil, err
+	}
+
+	dataSize := int64(v.Len()) * int64(v.Type().Elem().Size())
+	dataPtr := v.Pointer()
+
+	return newTensor(dataType, shape, unsafe.Pointer(dataPtr), dataSize)
+}
+
+// NewTensor creates a new Tensor that contains the specified data. The data type
+// and shape is deduced from the data parameter.
+// ex:
+//  - NewTensor("hello") // Creates scalar Tensor of type DTString
+//  - NewTensor([]int32{1, 2, 3}) // Creates a 1-D Tensor of type DTInt32
+//  - NewTensor([][]float32{{1, 2}, {3, 4}}) // Creates a 2-D Tensor of type DTFloat
```

Comment:
The idea is to support all the data types defined [here](https://github.com/tensorflow/tensorflow/pull/1771/files/562a1513b38aa29b222464ae19dd1ccbd0f48bfb#diff-6ecada9bdae40f6b8d2258c71de75c99R67) . Perhaps it would be a good idea to remove the unsupported data types. Do you think that is still necessary to specify the list here if I remove the unsupported there?